### DISCUSSION TITLE
Call to Api

### DISCUSSION THREAD

**Discussion by Lucho1976** [ipxe/ipxe#1053] [Q&A]
Hi everyone,

I'm trying to make an API call to my server to determine whether to boot from the disk or the network after a server has had Windows installed and it reboots.

I envision something like the code below (I will handle the platform later with another conditional). I tried using chain to make the call; but it shows "Ok" stops there. I also tried imgfetch, but I can't see the response. Although I can save the response to a file, I can't handle its content. As a workaround, I thought about sending a response code of 200 or 401 and using it in a conditional, and that works, but I'm unsure if the error is with the call or if the server isn't installed (see the example below).

Lastly, I can imagine making the API call from my DHCP server and returning different boot.ipxe files, but I haven't found any information on how to do that yet. Also i don't know if there is a solution in ipxe like the abortpxe.com in gpxe

This is what I envision:
```
#!ipxe

#! Attempt to chain to the script provided by your web server
chain http://myserver.com/server/28/installed || goto install_windows

#! If the server responds with "true", attempt to boot from the hard drive or just exit to continue with the next boot device
sanboot --no-describe --drive 0x80

exit

:install_windows
#! If the server responds with "false", proceed with the installation

kernel tftp://my.server.com/images/win2019/wimboot/wimboot
initrd tftp://my.server.com/images/win2019/boot/bcd bcd
initrd tftp://my.server.com/images/win2019/boot/boot.sdi boot.sdi
initrd tftp://my.server.com/images/win2019/sources/WinPE.wim boot.wim
boot
``` 

I tried this, and it works if I send a 200 or 401 response, but I can't differentiate between "not installed" and a "call error":

``` 
#!ipxe

imgfetch http://myserver.com && set response "true" || set response "false"
``` 

Thank you in advance!

**Comment by NiKiZe** [comment]
iPXE has support for script files.
Script files can be loaded over http
Many http implementations can generate contents dynamically.

Just make sure that what the server returns is a valid iPXE script.
Use ingfetch and imgstat to verify that what you get is detected as script.

**Reply by Lucho1976** [reply]
Hello, NiKiZe, thank you very much for your response; it was very helpful. I was only able to test the complete solution yesterday, and I'm encountering the following issue. When I make a chain call to a server to check if my OS is already installed, the PHP returns an iPXE script that sets an is_installed variable and goes into one label or another. This part works correctly. However, when it enters the branch where it has to load from the network, it loads the files, but when it tries to start the installation, I get the following error message:

Bad CPIO Magic
FATAL: could not extract initrd files

![image](https://github.com/ipxe/ipxe/assets/26300768/ff27bf26-5d13-49b8-8ec3-f4f167c397ab)

This is the ipxe Script:
```
#!ipxe

# Attempt to fetch the script from the PHP server
chain http://192.168.0.180/index.php?mac=${net0/mac}              

echo is_installed: ${is_installed}

# Check the value of 'is_installed' and boot from disk or network
iseq ${is_installed} 1 && goto installed || goto not_installed

:installed
echo Booting from disk.
exit

:not_installed
echo Installing from network.
set server_ip  192.168.0.180
kernel tftp://${server_ip}/images/win2019/wimboot/wimboot
initrd tftp://${server_ip}/images/win2019/boot/bcd bcd
initrd tftp://${server_ip}/images/win2019/boot/boot.sdi boot.sdi
initrd tftp://${server_ip}/images/win2019/sources/WinPE.wim boot.wim
boot
```
And this is the test php:
```
<?php

// Gets the MAC address from the GET parameters
$mac_address = $_GET['mac'];

// Define a test array of installed MAC addresses
$installed_macs = array("08:00:27:85:60:4axx");

// Checks if the MAC address is in the array
$is_installed = in_array($mac_address, $installed_macs);

// Generates the iPXE script to set the 'is_installed' variable
echo "#!ipxe\n";
echo "set is_installed $is_installed\n";
?>
```
But if I only include in my ipxe script the lines from the :not_installed tag, it works correctly. It seems as if the chain command gives it a different behavior. Do you know what might be causing this? I really appreciate your help

```
#!ipxe

echo Installing from network.
set server_ip  192.168.0.180
kernel tftp://${server_ip}/images/win2019/wimboot/wimboot
initrd tftp://${server_ip}/images/win2019/boot/bcd bcd
initrd tftp://${server_ip}/images/win2019/boot/boot.sdi boot.sdi
initrd tftp://${server_ip}/images/win2019/sources/WinPE.wim boot.wim
boot
```

**Reply by Lucho1976** [reply]
I tryied --autofree and worked. Is this ok?

chain --autofree http://192.168.0.180/index.php?mac=${net0/mac}

**Reply by NiKiZe** [reply]
Yes that is what you normally need.
